src/sage/rings/padics/padic_valuation.py

```diff
@@ -0,0 +1,1319 @@
+            raise NotImplementedError("can not rewrite %r which is defined on %r as a pseudo-valuation on %r"%(v, v.domain(), G.parent()))
+            
+
+        assert(v.domain() is G.parent())
+
+        # To obtain uniqueness of p-adic valuations, we need a canonical
+        # description of v. We consider all extensions of vK to L and select
+        # the one approximated by v.
+        vK = v.restriction(v.domain().base_ring()).extension(K)
+        if approximants is None:
+            approximants = vK.mac_lane_approximants(G)
+        approximants = [approximant.extension(v.domain()) for approximant in approximants]
+        approximant = vK.mac_lane_approximant(G, v, approximants=tuple(approximants))
+
+        return (R, approximant, L.construction()), {'approximants': approximants}
+
+    def create_key_and_extra_args_for_number_field_from_ideal(self, R, I, prime):
+        r"""
+        Create a unique key identifying the valuation of ``R`` with respect to
+        ``I``.
+
+        .. NOTE::
+
+            ``prime``, the original parameter that was passed to
+            :meth:`create_key_and_extra_args``, is only used to provide more
+            meaningful error messages
+
+        EXAMPLES::
+
+            sage: GaussianIntegers().valuation(GaussianIntegers().ideal(2)) # indirect doctest
+            2-adic valuation
+
+        """
+        K, L, G = self._normalize_number_field_data(R)
+
+        # To obtain uniqueness of p-adic valuations, we need a canonical
+        # description of v. We consider all extensions of vK to L and select
+        # the one approximated by v.
+        # Of course, this only works if I comes from a single prime downstairs.
+        p = I.relative_norm()
+        F = p.factor()
+        if len(F) != 1:
+            raise ValueError("%r does not lie over a single prime of %r"%(I, K))
+        vK = K.valuation(F[0][0])
+        candidates = vK.mac_lane_approximants(G)
+
+        candidates_for_I = [c for c in candidates if all(c(g.polynomial()) > 0 for g in I.gens())]
+        assert(len(candidates_for_I) > 0) # This should not be possible, unless I contains a unit
+        if len(candidates_for_I) > 1:
+            raise ValueError("%s does not single out a unique extension of %s to %s"%(prime, vK, L))
+        else:
+            # equality of number fields has it quirks since it says that two
+            # fields are == even if they are distinguishable (because they come
+            # from different constructions.)
+            # Including structure() into the key seems to be a way to distinguish such cases properly.
+            # This used to be an issue but seems to be fixed, namely, the
+            # absolute_field of a number field was deemed equivalent to the
+            # directly created absolute field, even though the absolute_field
+            # carried the information where it came from
+            return (R, candidates_for_I[0], L.construction()), {'approximants': candidates}
```

Comment:
Will L.construction() distinguish number fields recursively (e.g. for relative number fields)?